### ISSUE TITLE
vtable symbols are not marked artificial in DWARF

### ISSUE THREAD

**Issue by tromey** [rust-lang/rust#125126]
I tried this code, from the gdb testsuite:

```rust
fn empty () {
}

fn main () {
    let array = [1,2,3,4];
    let slice = &array[1..2];
    let hello = "hello";

    empty();			// STOP
}
```

I compiled with `-g`.

Examining the resulting DWARF, I see:

```
 <1><2a>: Abbrev Number: 2 (DW_TAG_variable)
    <2b>   DW_AT_name        : (indirect string, offset: 0xf3): <std::rt::lang_start::{closure_env#0}<()> as core::ops::function::Fn<()>>::{vtable}
    <2f>   DW_AT_type        : <0x3d>
    <33>   DW_AT_location    : 9 byte block: 3 80 24 5 0 0 0 0 0 	(DW_OP_addr: 52480)
```

I think this should be marked with `DW_AT_artificial`, as it does not correspond to any user-written code in the CU.  This will let gdb filter out these variables when displaying globals from a CU.

**Comment by cuviper** [comment]
We have `tests/codegen/debug-vtable.rs` making sure that the _type_ is artificial, and I also see that your example has `DIFlagArtificial` in the LLVM IR:

```llvm
!0 = !DIGlobalVariableExpression(var: !1, expr: !DIExpression())
!1 = distinct !DIGlobalVariable(name: "<std::rt::lang_start::{closure_env#0}<()> as core::ops::function::Fn<()>>::{vtable}", scope: null, file: !2, type: !3, isLocal: true, isDefinition: true)
!2 = !DIFile(filename: "<unknown>", directory: "")
!3 = !DICompositeType(tag: DW_TAG_structure_type, name: "<std::rt::lang_start::{closure_env#0}<()> as core::ops::function::Fn<()>>::{vtable_type}", file: !2, size: 384, align: 64, flags: DIFlagArtificial, elements: !4, vtableHolder: !14, templateParams: !23, identifier: "83627f9f9db11820e9b7128180480645")
```

But there's no sign of that in the output DWARF, neither on the variable nor its type:

```
 <1><2a>: Abbrev Number: 2 (DW_TAG_variable)
    <2b>   DW_AT_name        : (indirect string, offset: 0x7f): <std::rt::lang_start::{closure_env#0}<()> as core::ops::function::Fn<()>>::{vtable}
    <2f>   DW_AT_type        : <0x3d>
    <33>   DW_AT_location    : 9 byte block: 3 0 0 0 0 0 0 0 0 	(DW_OP_addr: 0)
 <1><3d>: Abbrev Number: 3 (DW_TAG_structure_type)
    <3e>   DW_AT_containing_type: <0xb5>
    <42>   DW_AT_name        : (indirect string, offset: 0x14d): <std::rt::lang_start::{closure_env#0}<()> as core::ops::function::Fn<()>>::{vtable_type}
    <46>   DW_AT_byte_size   : 48
    <47>   DW_AT_alignment   : 8
```

**Comment by cuviper** [comment]
For regular variables, the flag is applied [here](https://github.com/llvm/llvm-project/blob/fa9b1be45088dce1e4b602d451f118128b94237b/llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp#L1647-L1648), because that `Var.isArtificial()` checks [both the variable and the type](https://github.com/llvm/llvm-project/blob/fa9b1be45088dce1e4b602d451f118128b94237b/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h#L261-L268). I don't see anything like that for global varables, so I tried adding this:

```patch
--- a/llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp
+++ b/llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp
@@ -200,6 +200,9 @@ DIE *DwarfCompileUnit::getOrCreateGlobalVariableDIE(
   else
     addGlobalName(GV->getName(), *VariableDIE, DeclContext);

+  if (GTy && GTy->isArtificial())
+    addFlag(*VariableDIE, dwarf::DW_AT_artificial);
+
   addAnnotation(*VariableDIE, GV->getAnnotations());

   if (uint32_t AlignInBytes = GV->getAlignInBytes())
```

And now I do see it in DWARF:

```
 <1><2a>: Abbrev Number: 2 (DW_TAG_variable)
    <2b>   DW_AT_name        : (indirect string, offset: 0x63): <std::rt::lang_start::{closure_env#0}<()> as core::ops::function::Fn<()>>::{vtable}
    <2f>   DW_AT_type        : <0x3d>
    <33>   DW_AT_artificial  : 1
    <33>   DW_AT_location    : 9 byte block: 3 0 0 0 0 0 0 0 0 	(DW_OP_addr: 0)
```

I'll try an LLVM PR with this change soon.

**Comment by workingjubilee** [comment]
Is this a duplicate of https://github.com/rust-lang/rust/issues/90653 or?

I mean, I'd close that issue since this one is better enumerated, if they are, but

**Comment by cuviper** [comment]
That looks the same, yes.

**Comment by Enselic** [comment]
Triage: still waiting on https://github.com/llvm/llvm-project/pull/93118
